src/System.Text.Json/src/System/Text/Json/Serialization/ReadStackFrame.cs

```diff
@@ -11,76 +11,93 @@ namespace System.Text.Json.Serialization
+        public object ReturnValue;
+        public JsonClassInfo JsonClassInfo;
 
-        // Support Dictionary
-        internal string KeyName;
+        // Support Dictionary keys.
+        public string KeyName;
 
-        // Current property values
-        internal JsonPropertyInfo JsonPropertyInfo;
-        internal bool PopStackOnEndArray;
-        internal bool EnumerableCreated;
+        // Current property values.
+        public JsonPropertyInfo JsonPropertyInfo;
 
-        // Support System.Array and other types that don't implement IList
-        internal IList TempEnumerableValues;
+        // Pop the stack when the current array or dictionary is done.
+        public bool PopStackOnEnd;
+
+        // Support System.Array and other types that don't implement IList.
+        public IList TempEnumerableValues;
+        public bool EnumerableCreated;
 
         // For performance, we order the properties by the first deserialize and PropertyIndex helps find the right slot quicker.
-        internal int PropertyIndex;
-        internal List<PropertyRef> PropertyRefCache;
+        public int PropertyIndex;
+        public List<PropertyRef> PropertyRefCache;
 
-        // The current JSON data for a property does not match a given POCO, so ignore the property (recursively for enumerables or object).
-        internal bool Drain;
+        // The current JSON data for a property does not match a given POCO, so ignore the property (recursively).
+        public bool Drain;
 
-        internal void Initialize(Type type, JsonSerializerOptions options)
+        public void Initialize(Type type, JsonSerializerOptions options)
         {
             JsonClassInfo = options.GetOrAddClass(type);
+            InitializeJsonPropertyInfo();
+        }
+
+        public void InitializeJsonPropertyInfo()
+        {
             if (JsonClassInfo.ClassType == ClassType.Value || JsonClassInfo.ClassType == ClassType.Enumerable || JsonClassInfo.ClassType == ClassType.Dictionary)
             {
                 JsonPropertyInfo = JsonClassInfo.GetPolicyProperty();
             }
         }
 
-        internal void Reset()
+        public void Reset()
         {
-            ReturnValue = null;
+            Drain = false;
             JsonClassInfo = null;
+            KeyName = null;
             PropertyRefCache = null;
-            PropertyIndex = 0;
-            Drain = false;
-            ResetProperty();
+            ReturnValue = null;
+            EndObject();
         }
 
-        internal void ResetProperty()
+        public void ResetProperty()
         {
-            JsonPropertyInfo = null;
-            PopStackOnEndArray = false;
             EnumerableCreated = false;
+            JsonPropertyInfo = null;
+            PopStackOnEnd = false;
             TempEnumerableValues = null;
-            KeyName = null;
         }
 
-        internal bool IsProcessingEnumerableOrDictionary()
+        public void EndObject()
```

Comment:
Do we need this method? Where's it called other than from Reset()?

Reply:
It is consistent with the write side, so I'll leave it as self-documenting.